algorithms/integration/kapton_2019_correction.py

```diff
@@ -0,0 +1,538 @@
+
+    def get_kapton_path_mm(self, s1):
+        """ Get kapton path length traversed by an s1 vecto. If no kapton intersection or just touches the edge,
+            then None is returned"""
+        intersection_points = []
+        px=[]
+        # determine path length through kapton tape
+        for face in self.faces:
+            try:
+                px=map(int,face[0].get_ray_intersection_px(s1)) # faces are single panel anyway
+                if px[0] <0 or px[1] < 0 or px[0]>face[0].get_image_size()[0] or px[1]>face[0].get_image_size()[1]: continue
+                intersection_points.append(face[0].get_lab_coord(face[0].get_ray_intersection(s1))) # faces are single panel anyway
+            except Exception as e:
+                if 'DXTBX_ASSERT' not in str(e):
+                    print (str(e))
+
+        if len(intersection_points) == 0:
+            return 0.0
+        if len(intersection_points) ==1:
+            print ('Ray not intersecting 2 faces.Either s1 vector does not intersect kapton or just touches edge of kapton tape \
+                    No correction is needed for those cases.')
+            return 0.0
+
+        n_intersection_points=len(intersection_points)
+        kapton_path_mm=[]
+        for ii in range(n_intersection_points-1):
+            for jj in range(ii+1,n_intersection_points):
+                kapton_path_mm.append((col(intersection_points[ii])-col(intersection_points[jj])).length())
+        return max(kapton_path_mm)
+
+    def abs_correction(self, s1):
+        """ Compute absorption correction using beers law. Takes in a tuple for a single s1 vector and does absorption correction"""
+        kapton_path_mm=self.get_kapton_path_mm(s1)
+        if kapton_path_mm is not None:
+            absorption_correction = 1/math.exp(-self.abs_coeff * kapton_path_mm) # unitless, >=1
+        return absorption_correction
+
+    def abs_correction_flex(self, s1_flex):
+        """ Compute the absorption correction using beers law. Takes in a flex array of s1 vectors, determines path lengths for each
+            and then determines absorption correction for each s1 vector """
+        kapton_path_mm=flex.double()
+        kapton_faces=self.faces
+        from dials.algorithms.integration import get_kapton_path_cpp
+        # new style, much faster
+        kapton_path_mm=get_kapton_path_cpp(kapton_faces, s1_flex)
+        # old style, really slow
+        #for s1 in s1_flex:
+        #  kapton_path_mm.append(self.get_kapton_path_mm(s1))
+        # determine absorption correction
+        if kapton_path_mm is not None:
+            absorption_correction = 1/flex.exp(-self.abs_coeff * kapton_path_mm) # unitless, >=1
+        return absorption_correction
+
+    def zone_of_point_in_kapton_shadow(self,s1):
+        """Given an s1 vector, determine whether it is in the increasing/decreasing or no kapton shadow regime"""
+        pass
+
+    def distance_of_point_from_line(self, r0, r1, r2):
```

Comment:
This doesn't use `self`. If the maths are intrinsically linked to the Kapton class, then maybe it would be better as a static method, or otherwise a standalone function